### ISSUE TITLE
set pending breakpoint from within the extension

### ISSUE THREAD

**Issue by digittigerw** [Microsoft/MIEngine#530]
Hi there,

I'm writing a debugger extension based on MIEngine.  One thing I want to do is allow stop-at-main on debugger start, namely the debugger instead of user needs to auto set a breakpoint at main(). 

In MIDebugEngine/Engine.Impl/DebuggedProcess.cs->GetInitializeCommands(), there is this line:
    commands.Add(new LaunchCommand("-break-insert main", ignoreFailures: true));

It does set a break at main. But when the breakpoint gets hit, this code handles it:
    DebuggedProcess.HandleBreakModeEvent()
As this line
	AD7BoundBreakpoint[] bkpt = _breakpointManager.FindHitBreakpoints(bkptno, addr, frame, out fContinue);
fails to find the breakpoint in BreakpointManager, a continue command is issued later by this line
	//we hit a bp pending deletion
	//post the CmdContinueAsync operation so it does not happen until we have deleted all the pending deletes
	CmdContinueAsyncConditional(breakRequest);
That means the breakpoint is practically skipped.

To solve the problem, the best solution I can think of is to set the breakpoint at main by calling this
	Microsoft.MIDebugEngine.BreakpointManager.CreatePendingBreakpoint((IDebugBreakpointRequest2 pBPRequest, out IDebugPendingBreakpoint2 ppPendingBP)
or 
	Microsoft.MIDebugEngine.AD7Engine.CreatePendingBreakpoint(IDebugBreakpointRequest2 pBPRequest, out IDebugPendingBreakpoint2 ppPendingBP)

But the road block I run into is how to create the "pBRRequest" object for the API. My research so far yields no good way.

Could any one help me here ? 

Thanks !

**Comment by gregg-miskelly** [comment]
@digittigerw My guess is trying to go through all the AD7 infrastructure would be more painful then it is worth. Do you know why it doesn't go down this code path:

https://github.com/Microsoft/MIEngine/blob/8a63449cab405a88e76db58fdbe39252c13ba82d/src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs#L1029

That is the code that is trying to handle this scenario.

**Comment by digittigerw** [comment]
@gregg-miskelly I'm doing remote debugging, namely GDB talks to gdbserver. On first connection of gdb to gdbserver, the application is stopped at a point, probably its real entrypoint, which leads to execution of code 
https://github.com/Microsoft/MIEngine/blob/8a63449cab405a88e76db58fdbe39252c13ba82d/src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs#L991
that results in set of this.EntrypointHit, which in turns means line 1029 won't be reached when the breakpoint-at-main is hit.

Even if I hack the code so that "this.EntrypointHit" is not set at line 991, namely line 1029 will be reached,  this line "_callback.OnEntryPoint(thread);" will eventually lead to a Continue command by the framework, namely the break-at-main is skipped.  

So that route is not a good solution to pursue.
Thanks !

**Comment by gregg-miskelly** [comment]
@digittigerw So we can get the debugger to stop at entry point, but only if -- when LaunchDebugTargets4 is called, the [launch flags][https://msdn.microsoft.com/en-us/library/microsoft.visualstudio.shell.interop.vsdebugtargetinfo4.launchflags.aspx] include DBGLAUNCH_StopAtEntryPoint. Normally the way this works is that this will get set if the user starts debugging with F10/F11, but not get set with F5. However, that isn't going to work if the debug engine is already issuing an entry point event from the first fake entry point as that needs to get ignored.

One possible fix:
* Pass DBGLAUNCH_StopAtEntryPoint so that when the entry point is sent the UI will actually break
* Add a success handler to [this line](https://github.com/Microsoft/MIEngine/blob/8a63449cab405a88e76db58fdbe39252c13ba82d/src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs#L719) so that the MI Engine will understand the breakpoint number of the entry point breakpoint. Store this in like an `int? _entryPointBpNo`.
* Add a new launch option like 'IgnoreFirstBreakpoint' (maybe with a better name) that will cause the entry point code to ignore the initial stop
* Update the stopping event handler to do something like `if (_launchOptions.IgnoreFirstBreakpoint && _entryPointBpNo.HasValue && hitBpNo != _entryPointBpNo.Value) { continue debuggee }`. This will take care of ignoring that first stopping event.

**Comment by digittigerw** [comment]
@gregg-miskelly 
I see that F10/F11 would make the debugger stop on "entrypoint" event.
But as to the difference of F10/F11 from F5 in launching debugger, is it a published/documented behavior ? If not, that would be a hidden feature that is hard for users to know.  Also, what if user just uses the menu or toolbar button to start debugger ?

And technically, in MIEngine, how can I tell if DBGLAUNCH_StopAtEntryPoint is set so as do corresponding handling ?

BTW, I don't see DBGLAUNCH_StopAtEntryPoint in this doc
https://msdn.microsoft.com/en-us/library/microsoft.visualstudio.shell.interop.__vsdbglaunchflags.aspx
nor do I see it in source code of Microsoft.VisualStudio.Shell.Interop.__VSDBGLAUNCHFLAGS in VS 2015.

**Comment by gregg-miskelly** [comment]
RE: Documentation - I am certain there is some documentation for this behavior. Here is the first hit I found: https://blogs.msdn.microsoft.com/visualstudio/2015/05/22/debugging-tips-and-tricks/. Though I am not sure how much documentation there is has much effect on how many people know about the behavior :).

RE: Menu/Toolbar - Yes, they all act the same (this isn't a special feature just for the keyboard shortcut).

RE: How to set DBGLAUNCH_StopAtEntryPoint - are you creating a project system, or are you using the command window? If you are using the command window, then here is the code: https://github.com/Microsoft/MIEngine/blob/master/src/MIDebugPackage/MIDebugPackagePackage.cs#L362

RE: How can MIEngine tell if StopAtEntryPoint is set - it doesn't need to know. It just sends the entry point event. If the user has started debugging with step into/over, then the UI will stop at that event. Otherwise it will not.

RE: MSDN - yup, sadly MSDN documentation of VS interfaces isn't great. You can see it here: https://msdn.microsoft.com/en-us/library/microsoft.visualstudio.shell.interop.__vsdbglaunchflags2(vs.80).aspx

**Comment by digittigerw** [comment]
Thanks for all the info, Gregg!
> RE: Menu/Toolbar - Yes, they all act the same (this isn't a special feature just for the keyboard shortcut).
Do you mean this _is_ a special feature just for keyboard shortcut ?

**Comment by gregg-miskelly** [comment]
@digittigerw no. The menu item (Debug->Step Into), the toolbar icon, and the keyboard shortcut all run the same code.

**Comment by digittigerw** [comment]
Oh, I see. I was just thinking about the menu Debug->Start Debugging and its toolbar counterpart. I must say it's not  intuitive for user to click the "step over" or "step into" button to start the debugger :) But I got your point.
